pkg/rtc/transport_test.go

```diff
@@ -269,6 +271,70 @@ func TestNegotiationFailed(t *testing.T) {
 	}, negotiationFailedTimout+time.Second, 10*time.Millisecond, "negotiation failed")
 }
 
+func TestFilteringCandidates(t *testing.T) {
+	params := TransportParams{
+		ParticipantID:       "id",
+		ParticipantIdentity: "identity",
+		Target:              livekit.SignalTarget_PUBLISHER,
+		Config:              &WebRTCConfig{},
+		EnabledCodecs: []*livekit.Codec{
+			{Mime: webrtc.MimeTypeOpus},
+			{Mime: webrtc.MimeTypeVP8},
+			{Mime: webrtc.MimeTypeH264},
+		},
+	}
+	transport, err := NewPCTransport(params)
+	require.NoError(t, err)
+
+	_, err = transport.pc.CreateDataChannel("test", nil)
+	require.NoError(t, err)
+
+	_, err = transport.pc.AddTransceiverFromKind(webrtc.RTPCodecTypeAudio)
+	require.NoError(t, err)
+
+	_, err = transport.pc.AddTransceiverFromKind(webrtc.RTPCodecTypeVideo)
+	require.NoError(t, err)
+
+	offer, err := transport.pc.CreateOffer(nil)
+	require.NoError(t, err)
+
+	offerGatheringComplete := webrtc.GatheringCompletePromise(transport.pc)
+	require.NoError(t, transport.pc.SetLocalDescription(offer))
+	<-offerGatheringComplete
+
+	// should not filter out UDP candidates if TCP is not preferred
+	offer = *transport.pc.LocalDescription()
+	filteredOffer := transport.FilterCandidates(offer)
+	require.EqualValues(t, offer.SDP, filteredOffer.SDP)
+
+	parsed, err := offer.Unmarshal()
+	require.NoError(t, err)
+
+	getNumUDPCandidates := func(sdp *sdp.SessionDescription) int {
+		numUDPCandidates := 0
+		for _, a := range sdp.Attributes {
+			if a.Key == "candidate" && strings.Contains(a.Value, "udp") {
+				numUDPCandidates++
+			}
+		}
+		for _, m := range sdp.MediaDescriptions {
+			for _, a := range m.Attributes {
+				if a.Key == "candidate" && strings.Contains(a.Value, "udp") {
+					numUDPCandidates++
+				}
+			}
+		}
+		return numUDPCandidates
+	}
+	require.NotZero(t, getNumUDPCandidates(parsed))
+
+	transport.SetPreferTCP(true)
+	filteredOffer = transport.FilterCandidates(offer)
+	parsed, err = filteredOffer.Unmarshal()
+	require.NoError(t, err)
+	require.Zero(t, getNumUDPCandidates(parsed))
```

Comment:
Wanted to actually add TCP candidates and ensure that it is not filtered. But, seems like setting that up in a test is a bit laborious. Maybe, I can just add a couple of hard coded TCP candidates?

Reply:
sg for the test.

Reply:
Actually, went ahead and added canned candidates and testing with it. Just to be safe.